docs/architecture/adr-046-module-params.md

```diff
@@ -0,0 +1,165 @@
+module parameters. Namely, managing parameters essentially comes for "free" in
+that developers only need to define the `Params` struct, the `Subspace`, and the
+various auxiliary functions, e.g. `ParamSetPairs`, on the `Params` type. However,
+there are some notable drawbacks. These drawbacks include the fact that parameters
+are serialized in state via JSON which is extremely slow. In addition, parameter
+changes via `ParamChangeProposal` governance proposals are _stateless_. In other
+words, it is currently not possible to have any state transitions in the
+application during an attempt to change param(s).
+
+## Decision
+
+We will build off of the alignment of `x/gov` and `x/authz` work per
+[#9810](https://github.com/cosmos/cosmos-sdk/pull/9810). Namely, module developers
+will create one or more unique parameter data structures that must be serialized
+to state. In addition, modules must implement `sdk.Msg` message(s) and their
+respective handler(s) such that when a governance proposal passes, via the work
+done in [#9810](https://github.com/cosmos/cosmos-sdk/pull/9810), the `x/gov`
+module can execute those messages which would then change the respective
+parameters and any other state mutations necessary.
+
+Note, it is up to developers to decide how to structure their parameters and
+the respective `sdk.Msg` messages. Consider the parameters currently defined in
+`x/auth` using the `x/params` module for parameter management:
+
+```protobuf
+message Params {
+  uint64 max_memo_characters       = 1;
+  uint64 tx_sig_limit              = 2;
+  uint64 tx_size_cost_per_byte     = 3;
+  uint64 sig_verify_cost_ed25519   = 4;
+  uint64 sig_verify_cost_secp256k1 = 5;
+}
+```
+
+Developers can choose to either create a unique data structure for every field in
+`Params` or they can create a single `Params` structure as outlined above in the
+case of `x/auth`.
+
+In the former approach, a `sdk.Msg` would need to be created for every single
+field along with a handler. This can become burdensome if there are a lot of
+parameter fields. In the latter case, there is only a single data structure and
+thus only a single message handler, however, the message handler might have to be
+more sophisticated in that it might need to understand what parameters are being
+changed vs what parameters are untouched.
+
+Once the `sdk.Msg` types and corresponding handlers are defined, proposals can
+be made using the `x/gov` module that are authorized by the root `x/gov`
+module's account.
+
+Continuing to use `x/auth`, we demonstrate a more complete example:
+
+```go
+type Params struct {
+	MaxMemoCharacters      uint64
+	TxSigLimit             uint64
+	TxSizeCostPerByte      uint64
+	SigVerifyCostED25519   uint64
+	SigVerifyCostSecp256k1 uint64
+}
+
+type MsgUpdateParams struct {
+	MaxMemoCharacters      uint64
+	TxSigLimit             uint64
+	TxSizeCostPerByte      uint64
+	SigVerifyCostED25519   uint64
+	SigVerifyCostSecp256k1 uint64
+}
```

Comment:
The types above are the same. No need to duplicate.

Reply:
We could use a convention that the method to update a params = params struct. So, below we would have:
```
func (ms msgServer) Params(goCtx context.Context, msg *types.MsgParams) (*types.MsgParamsResponse, error)
```